### ISSUE TITLE
Obtaining native window handle

### ISSUE THREAD

**Issue by Xeverous** [mosra/magnum#479]
I'm creating an application that (so far) is planned to be run on desktop (Linux and Windows) and the browser (via Emscripten). I need to ask the user to open a file - I'm not sure yet how such thing will be done in Emscripten build, but for Windows (or any other desktop OS) I need the native window handle.

There is `SDL_Window* Magnum::Platform::Sdl2Application::window()` and https://stackoverflow.com/questions/24117983/get-window-handle-of-sdl-2-application guides to this snippet:

```cpp
SDL_SysWMinfo wmInfo;
SDL_VERSION(&wmInfo.version);
SDL_GetWindowWMInfo(window, &wmInfo);
HWND hwnd = wmInfo.info.win.window;
```

... so obtaining native window handle (at least on SDL-based applcation) is not a problem.

The questions are:

- What would be a correct place (in code) to spawn modal window? On Windows (`IFileSaveDialog::Show(hwnd)`), if the modal window is given the pointer to the parent window, the parent window is blocked and **event handling stops there untill modal is closed because the modal temporarily takes ownership of the OS event loop** - I had problems with other libraries where doing such things (spawning modal windows) inside event handling function could cause the application to stop, loop infinitely or break in some other way (any sort of unfixable event loop corruption). Usually I had to just delay spawning the modal to a different application function, outside event callbacks.
- Is there an inherent limitation that `Platform::*Application` can only create 1 window?
- Is there any better way to do it in Magnum? I haven't found any explicit API for OS modal windows

**Comment by mosra** [comment]
Hi!

> Is there an inherent limitation that Platform::*Application can only create 1 window?

Currently, yes. There are various in-progress but stale attempts such as #168, however due to focus being elsewhere I didn't have time to make any progress there.

What people end up doing for multi-window apps instead is sidestepping the `Application` classes and using SDL (GLFW, ...) directly. But to be frank I don't know how modal windows are done in SDL/GLFW either, I only ever used them in Qt, and there it was as you describe -- the event loop got paused for the main window and the modal window got the input instead.

Hmm, I guess your *real* question is how to open a native file open/save dialog, right? That's definitely something that would be nice to have.

Getting native window handles in SDL/GLFW could be implemented in the `Application` classes, but I guess that alone wouldn't be as useful if you would have no working way to open dialogs from those, right?

**Comment by mosra** [comment]
> at this point there is no incentive to support legacy dialogs - the new ones are available in Vista

Yeah I don't intentionally support XP either anyway, so this should be fine.

> wchar

Yep, there's [Utility::Unicode::widen()](https://doc.magnum.graphics/corrade/namespaceCorrade_1_1Utility_1_1Unicode.html#af92cf003d4c9bf36b0a53777ab5a0445), which wraps the Windows APIs (and thus is Windows-specific because no other platform really needs this anyway) and which is used by all Magnum APIs that deal with Unicode on Windows.

> tinyfiledialogs

That's exactly what I wanted to base the implementation on / steal from -- doing some `dlopen()` in case the API isn't implicitly linked from elsewhere.

> UI on Linux

I'm aware :) I'd go with KDE and GTK support. With KDE it'd be I assme rather easy, unfortunately [as far as I know](https://www.reddit.com/r/gamedev/comments/2njlss/i_just_open_sourced_a_tiny_c_api_w_minimal_bloat/cmebd64/?utm_source=reddit&utm_medium=web2x&context=3) with GTK there's no way to unload the library so once it gets loaded it'll be in memory forever. I think because of this SDL had to use some sort of IPC to avoid loading GTK in [SDL_ShowMessageBox](https://wiki.libsdl.org/SDL_ShowMessageBox) (can't find the source tho).

> Give a commit link how I have integrated tinyfiledialogs into my project

Yeah that would be great. But please note that I'm still not able to invest any time into this myself in the near future.

**Comment by Xeverous** [comment]
> That's exactly what I wanted to base the implementation on / steal from -- doing some `dlopen()` in case the API isn't implicitly linked from elsewhere.

Then tinyfiledialogs should be the prefferred choice. I have searched for many similar solutions (found also the one you linked) and this one has least dependencies, is completely troubleless, no linking required (except on Windows to some system DLLs), no global state init required, fully C and C++ compliant, single file implementation ... basically ideal.

> with GTK there's no way to unload the library so once it gets loaded it'll be in memory forever

Ahh, I remember similar nightmare trying to RAII-ize fontconfig library. It has one of the worst C library designs I have seen when it comes to initialization:

- There are 3 different initialization ways; one depends on reading external XML file.
- There is no way to check if library has been initialized.
- Any extra initialization attempt is a no-op.
- Any extra cleanup attemt is undefined behavior.

A single project can deal with it. But when your project has 2 dependencies and both of them try to correctly use fontconfig, it is a distaster. The only way to "fix" it is to force a leak and make sure fontconfig is never attempted to be cleaned up twice. I just can't get why it is so complex, why there is no init check and why extra init/cleanup usage is inconsistent. Many libraries have just an integer reference counter or simply create 2 independent instances.

>> Give a commit link how I have integrated tinyfiledialogs into my project
>
> Yeah that would be great. But please note that I'm still not able to invest any time into this myself in the near future.

Fine, I already see you must be busy by looking at the reponse times on this issue. I'm currently more involved in using ImGui than Magnum.

https://github.com/Xeverous/filter_spirit/commit/3e06e81a40c9655fddbb8d1a9cefbd6fb81ed5e4

This commit is actually bigger than I thought (it also adds stuff to my project that immediately uses it) so I think a more comfortable way would be to state that everything is in this path: https://github.com/Xeverous/filter_spirit/tree/3e06e81a40c9655fddbb8d1a9cefbd6fb81ed5e4/external/tiny_file_dialogs - the whole thing is just a single CMake file + subproject link (link is not clickable because tinyfiledialogs project is not located on github).
